src/NuGet.Core/NuGet.Protocol/Resources/AutoCompleteResourceV3.cs

```diff
@@ -51,8 +51,8 @@ public AutoCompleteResourceV3(HttpSource client, ServiceIndexResourceV3 serviceI
 
             var queryUri = queryUrl.Uri;
             var results = await _client.GetJObjectAsync(
-                new HttpSourceRequest(queryUri, Common.NullLogger.Instance),
```

Comment:
Thank you for your contribution.
Maybe `ILogger  logger = log ?? Common.NullLogger.Instance;` then reusing it might look better. Also adding asserting logger log in following places would be great idea.
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Protocol.Tests/AutoCompleteResourceV3Tests.cs
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Protocol.Tests/PowershellAutoCompleteResourceTests.cs

Reply:
Thanks, will do!

Reply:
Sorry, @erdembayar - quick question:

> Also adding asserting logger log in following places would be great idea.

If I understand this correctly, the test should assert that the supplied `ILogger` instance is used, and not `NullLogger.Instance`?

`NullLogger` is a singleton (and there doesn't seem to be any other test-friendly implementation of `ILogger` that I can easily create an instance of, in order to verify that the supplied `ILogger` instance is used rather than `NullLogger.Instance`.

Off the top of my head, options for this are:

1. Use Moq's `Mock<T>` to mock out an `ILogger` (potentially tricky because I don't know what calls downstream components will make to it so it might make the test a little fragile)
2. Create a new logger class (derived from `LoggerBase`) just for use in these tests
3. Just try to get a separate instance of `NullLogger` to play with (eg. via `Activator.CreateInstance` or similar)?

Reply:
I would go with 2nd option. You can create new logger  class implements ILogger and pass to [resource](https://github.com/NuGet/NuGet.Client/blob/f93c75477499958533a1b3ade387a20708ba8fac/test/NuGet.Core.Tests/NuGet.Protocol.Tests/PowershellAutoCompleteResourceTests.cs#L107) and check if anything is logged into logger after operation. 
If you see anything in logged then it's good.

Reply:
> You can create new logger class implements ILogger

We have a TestLogger. 

Please uste that.

Reply:
Ok, I've added a logging check to the test for `IdStartsWithAsync`, but there isn't a test for `VersionStartsWithAsync`. If you're happy with the changes I made to the first test I'm happy to have a go at writing the second one.

Reply:
Please check my [comment](https://github.com/NuGet/NuGet.Client/pull/4287#discussion_r720556123).